Added Contributing Guidelines

estermer opened the pull request:
#116 

In this file, I copied the guidelines found in blockstack-browser [contributing](https://github.com/blockstack/blockstack-browser/blob/master/CONTRIBUTING.md) and edited them to match the needs of the `blockstack.org` repo.

Items changed:

- urls to repo in a few spots such as labels and steps to clone the repo
- changed the list of labels to match those found in [.orgs labels](https://github.com/blockstack/blockstack.org/labels)
- edited the process to create a PR (name of branches and what base should be)

This should be a good start. If there are any further requests, let me know.

guylepage3 commented:
This is awesome @estermer! 👍  Will take a look over the weekend. This will assist the community immensely as it will save everyone so much time.

estermer commented:
@guylepage3 @shea256 I added you both as maintainers to the bottom of the guidelines and the branch your in charge of. Hopefully this will help also with community members on who to contact for discussion and who to expect to be conversing with about PR's.

estermer commented:
@guylepage3 after reviewing the flow of the contrib guidelines, I went and added a few features that I thought would be nice.

Including:
- Table of Contents in README to make the contrib guide link priority
- added back to the top links at bottom of each section to make navigation on the page easier
- added a Finding opportunities to contribute as the top section to encourage contributors how to get started.

I also fixed a few issues with linkage to some external resources and relative to repo verbiage.

This should be good for a review from you whenever you have some time.

guylepage3 commented:
Awesome @estermer. I'll take a look this afternoon. Great work! 😄

guylepage3 commented:
@estermer I've moved this PR into the "For Review" column here: https://github.com/blockstack/blockstack.org/projects/2

guylepage3 commented:
commented here: https://github.com/blockstack/blockstack.org/issues/322#issuecomment-311749995

guylepage3 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,272 @@
+      - master
+   Branch name for "next release" development: [develop] develop
+   
+   How to name your supporting branch prefixes?
```
I like the `git flow` process but i feel as though they go a bit too far by adding the `issue number` and `github username`. This seems to be a bit much to execute a pull request.

Thoughts @larrysalibra @vsund ?

guylepage3 replied:
cc @estermer

vsund replied:
We could frame this as "if you try to do the perfect pull request, you can do [...]" but don't require this for every pull request (especially because there will be pull requests that fix just a small typo) We could try to handle such pull requests separately with a smaller guide though.

This also applies to the usage of `git flow` at all :)

estermer replied:
I believe @larrysalibra was the one who suggested the issue number because of githubs ability to reference, I think we could do without it. It would be nice to just doe feature name and github username.

vsund replied:
I think GitHub only recognizes them in commit messages and other text fields (if we want this for automatic referencing).

guylepage3 replied:
>feature name and github username

I personally feel we can go one step further with simplification and drop the username as there will be reference to the contributor in the commit/pull request itself.

guylepage3 replied:
So having said that, @estermer can you please remove the `issue-` references in the `git flow` setup, `<issue number>-` + `-<github username>` in the command example and any other references to these two added items throughout the `CONTRIBUTING.md` file?

Other than the items above, this is absolutely brilliant. 😄

guylepage3 replied:
So having said that, @estermer can you please remove the `issue-` references in the `git flow` setup, `<issue number>-` + `-<github username>` in the command example and any other references to these two added items throughout the `CONTRIBUTING.md` file?

Other than the items above, this is absolutely brilliant. 😄

guylepage3 replied:
also cc;ing @shea256 @jackzampolin for feedback

jackzampolin replied:
I think theres a case for keeping the issue # if applicable, but because we are asking developers to fork the project and contribute the `username` section is going to be superfluous. 

While I will agree with @vsund that it would be nice to have a separate process for one line fixes, I believe there is significant value in having a standardized process for all pulls. Therefore I think if we are going to require it for some we should require it for all.

larrysalibra replied:
To be clear, git flow only needs to be used by developers that have commit access to the repository. For someone looking to contribute, they shouldn't need to use get flow. All contributors need to do is to make sure they base their pull requests on either the `develop` or the specific feature branch they're trying to contribute to.

guylepage3 replied:
>All contributors need to do is to make sure they base their pull requests on either the develop or the specific feature branch they're trying to contribute to.

They should also be creating their own feature branches so that the maintainer of the repo does not have to do so.

guylepage3 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,272 @@
+
+[^ Back To The Top](#contributing-to-blockstackorg)
+
+## Maintainers
```
@estermer some in the community have pointed out that we should tone down the "Maintainers" section of the guide..

Why don't we go with something like this...

|  Maintainer        |  GitHub  |
|  --------------  |  ----------------------------------------------  |
|  Ryan Shea        | [shea256](https://github.com/shea256)             |
|  Guy Lepage      | [guylepage3](https://github.com/guylepage3)  |

estermer replied:
change made

larrysalibra commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,272 @@
+
+[^ Back To The Top](#contributing-to-blockstackorg)
+
+## Bug reports
```
Some of this info duplicates the "when reporting a bug section above"

estermer replied:
duplicates removed

larrysalibra commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,272 @@
+
+### JS
+
+- No semicolons (in client-side JS)
```
Let's remove this and define the javascript requirements programmatically with the `.eslintrc` file instead of english.

estermer replied:
do we want a link to the eslint repo for their .eslintrc file?

guylepage3 replied:
@larrysalibra, thoughts?

larrysalibra commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,272 @@
+- "Attractive"
+- Don't use jQuery (no "$" allowed)
+
+### Checking code
```
Should be `npm run test` (or equivalent for other repos)

estermer replied:
i think this is just `gulp test` on this repo

larrysalibra commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,272 @@
+
+[^ Back To The Top](#contributing-to-blockstackorg)
+
+## Pull requests
```
Pull requests should have tests for whatever functionality they're changing or implementing.

estermer replied:
there is a spot that covers this, I am adding your suggestion on making a short and sweet contribution guide which includes a couple musts with writing tests being one of them.

guylepage3 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,272 @@
+
+* Reduced test cases and potential fixes using [JS Bin](https://jsbin.com)
+
+### Issues and labels
```
Very good point @larrysalibra. Let's remove this section altogether.

estermer commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,272 @@
+
+* Please **do not** derail or troll issues. Keep the discussion on topic and respect the opinions of others.
+
+* Please **do not** post comments consisting solely of "+1" or ":thumbsup:". Use [GitHub's "reactions" feature](https://github.com/blog/2119-add-reactions-to-pull-requests-issues-and-comments) instead. We reserve the right to delete comments which violate this rule.
```
we could leave it in so contributors can at least work try and follow it, I can remove the right to delete comments that do this and just not enforce this. But again I think it is nice to have because it is a cleaner way to show support rather than scrolling through a page of +1's

guylepage3 replied:
I'm in agreement with @larrysalibra on this one @estermer. Reason being is that the friction level of reading this is much greater than the friction of scrolling through +1's. If it gets out of control, then we an add this back in.

guylepage3 replied:
I'm in agreement with @larrysalibra on this one @estermer. Reason being is that the friction level of reading this is much greater than the friction of scrolling through +1's. If it gets out of control, then we an add this back in.

estermer commented:
OK I added TL;DR to the README.md to shorten everything up and linked to the guide incase they decide to read through. I removed the labels and the do not +1, as well as the git flow steps in the PR process since that is only for those who have commit access. I deduped the bug report info in the guide. Also linked to ESLint as our JS style. I also toned down the maintainer section to just the name and link.
